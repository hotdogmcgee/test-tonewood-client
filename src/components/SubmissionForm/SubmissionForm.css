.SubmissionPage {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.Input,
.Textarea {
  margin-bottom: 0px;
}

.Input {
  text-align: right;
}

.Input::placeholder,
.Textarea::placeholder {
  color: grey;
}
*/ form {
  max-width: 1000px;
  display: flex;
  flex-direction: column;
}

form > div {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  max-width: 400px;
  margin-bottom: 10px;
}

.comments {
  height: 150px;
  align-items: inherit;
}

.comments textarea {
  text-align: left;
}

select {
  width: 55%
}
