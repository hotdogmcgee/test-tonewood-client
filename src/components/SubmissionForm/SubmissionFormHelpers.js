const mathConst = 0.94625;
const milli = 0.001;
const nano = Math.pow(10, -9);
const giga = Math.pow(10, 9);

const Formulas = {
  getVolume(length, width, height) {
    return length * milli * (width * milli) * (height * milli);
  },

  getDensity(length, width, height, mass) {
    const volume = this.getVolume(length, width, height);
    const kgMass = mass * milli;
    return (kgMass / volume).toFixed(2);
  },

  getELong(length, width, height, mass, peakFreqLong) {
    const density = this.getDensity(length, width, height, mass);
    const eLong =
      ((mathConst *
        density *
        Math.pow(length * milli, 4) *
        Math.pow(peakFreqLong, 2)) /
        Math.pow(height * milli, 2)) *
      nano;
    return eLong.toFixed(2);
  },

  getECross(length, width, height, mass, peakFreqCross) {
    const density = this.getDensity(length, width, height, mass);
    const eCross =
      ((mathConst *
        density *
        Math.pow(width * milli, 4) *
        Math.pow(peakFreqCross, 2)) /
        Math.pow(height * milli, 2)) *
      nano;
    return eCross.toFixed(2);
  },

  getVelocitySoundLong(length, width, height, mass, eLong) {
    const density = this.getDensity(length, width, height, mass);
    const velocitySoundLong = Math.sqrt((eLong * giga) / density) * milli;
    return velocitySoundLong.toFixed(2);
  },

  getRadiationRatio(length, width, height, mass, velocitySoundLong) {
    const density = this.getDensity(length, width, height, mass);
    const radiationRatio = (velocitySoundLong / density) * 1000;
    return radiationRatio.toFixed(2);
  }
};

export default Formulas;
