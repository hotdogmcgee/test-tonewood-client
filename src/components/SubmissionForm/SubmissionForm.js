import React from "react";
import WoodApiService from "../../services/wood-api-service";
import { Button, Textarea, NumericFormFields } from "../Utils/Utils";
import WoodListContext from "../../contexts/WoodListContext";
import "./SubmissionForm.css";

export default class SubmissionForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      optionValue: "none",
      error: null
    };
    this.handleChange = this.handleChange.bind(this);
  }

  static defaultProps = {
    onSubmissionSuccess: () => {}
  };

  static contextType = WoodListContext;

  componentDidMount() {
    this.context.clearError();
    WoodApiService.getWoods()
      .then(this.context.setWoodsList)
      .catch(this.context.setError);
  }

  handleSubmit = e => {
    e.preventDefault();
    const {
      tw_id,
      new_tw_name,
      density,
      e_long,
      e_cross,
      velocity_sound_long,
      radiation_ratio,
      sample_length,
      sample_width,
      sample_thickness,
      sample_weight_grams,
      peak_hz_long_grain,
      peak_hz_cross_grain,
      comments
    } = e.target;
    const items = [
      tw_id,
      new_tw_name,
      density,
      e_long,
      e_cross,
      velocity_sound_long,
      radiation_ratio,
      sample_length,
      sample_width,
      sample_thickness,
      sample_weight_grams,
      peak_hz_long_grain,
      peak_hz_cross_grain,
      comments
    ];
    WoodApiService.postSubmission({
      tw_id: tw_id.value,
      new_tw_name: new_tw_name.value || null,
      density: density.value,
      e_long: e_long.value,
      e_cross: e_cross.value,
      velocity_sound_long: velocity_sound_long.value,
      radiation_ratio: radiation_ratio.value,
      sample_length: sample_length.value,
      sample_width: sample_width.value,
      sample_thickness: sample_thickness.value,
      sample_weight_grams: sample_weight_grams.value,
      peak_hz_long_grain: peak_hz_long_grain.value,
      peak_hz_cross_grain: peak_hz_cross_grain.value,
      comments: comments.value
    })
      .then(() => {
        for (let i = 0; i < items.length; i++) {
          items[i].value = "";
        }
        this.props.onSubmissionSuccess();
      })
      .catch(res => {
        this.context.setError();
      });
  };

  handleChange(event) {
    this.setState({ optionValue: event.target.value });
  }

  renderTwOptions() {
    const { woodsList = [] } = this.context;
    return woodsList.map(wood => {
      return (
        <option value={wood.id} key={wood.id}>
          {wood.common_name} ({wood.genus} {wood.species})
        </option>
      );
    });
  }

  setStyle() {
    if (this.state.optionValue !== "1") {
      return { display: "none" };
    }
  }

  renderSubmissionForm() {
    return (
      <form className="SubmissionForm" onSubmit={this.handleSubmit}>
        <div className="tw_id">
          <label htmlFor="tw_id">Select your tonewood</label>
          <select
            required
            aria-label="Select your tonewood!"
            name="tw_id"
            id="tonewood-select"
            className="tw_select"
            value={this.state.optionValue}
            onChange={this.handleChange}
          >
            <option value="none" disabled hidden>
              Select a Wood
            </option>
            {this.renderTwOptions()}
          </select>
        </div>

        <div style={this.setStyle()} className="new_tw_name">
          <label htmlFor="SubmissionForm__new_tw_name" value="new_tw_name">
            New tonewood
          </label>
          <Textarea
            name="new_tw_name"
            className="Input_new_tw_name"
            placeholder="Specify scientific name if possible"
          ></Textarea>
        </div>

        <NumericFormFields />
        <div className="comments">
          <label htmlFor="SubmissionForm__comments">Comments</label>
          <Textarea
            name="comments"
            className="Input_comments"
            placeholder="Write something"
          ></Textarea>
        </div>

        <Button type="submit">Add submission</Button>
      </form>
    );
  }

  render() {
    return this.renderSubmissionForm();
  }
}
