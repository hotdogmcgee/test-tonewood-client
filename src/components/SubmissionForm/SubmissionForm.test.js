import React from "react";
import SubmissionForm from "./SubmissionForm";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

describe("SubmissionForm component", () => {
  it("renders a SubmissionForm by default", () => {
    const wrapper = shallow(<SubmissionForm />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});