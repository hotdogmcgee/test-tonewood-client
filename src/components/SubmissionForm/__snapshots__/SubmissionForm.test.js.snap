// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmissionForm component renders a SubmissionForm by default 1`] = `
<Fragment>
  <form
    className="SubmissionForm"
    onSubmit={[Function]}
  >
    <div
      className="tw_id"
    >
      <label
        htmlFor="tw_id"
      >
        Select your tonewood
      </label>
      <select
        aria-label="Select your tonewood!"
        className="tw_select"
        id="tonewood-select"
        name="tw_id"
        onChange={[Function]}
        required={true}
        value="none"
      >
        <option
          disabled={true}
          hidden={true}
          value="none"
        >
          Select a Wood
        </option>
        <option
          key="1"
          value="1"
        >
          Other
        </option>
      </select>
      <ValidationError
        className="tw_id_error"
        hasError={true}
        message=""
      />
    </div>
    <div
      id="new_tw_name"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <label
        htmlFor="SubmissionForm__new_tw_name"
        value="new_tw_name"
      >
        New tonewood. 
        <FontAwesomeIcon
          border={false}
          className="info-icon"
          fixedWidth={false}
          flip={null}
          icon="info-circle"
          inverse={false}
          listItem={false}
          mask={null}
          onClick={[Function]}
          pull={null}
          pulse={false}
          rotation={null}
          size="1x"
          spin={false}
          symbol={false}
          title=""
          transform={null}
        />
      </label>
      <Textarea
        id="Input_new_tw_name"
        name="new_tw_name"
        onChange={[Function]}
        placeholder="Specify scientific name."
      />
      <ValidationError
        hasError={false}
        message=""
      />
    </div>
    <NumericFormFields />
    <div
      className="comments"
    >
      <label
        htmlFor="SubmissionForm__comments"
      >
        Comments
      </label>
      <Textarea
        className="Input_comments"
        name="comments"
        placeholder="Write something"
      />
    </div>
    <Button
      disabled={true}
      type="submit"
    >
      Add submission
    </Button>
    <Section
      id="SubmitReady__Section"
    >
      <h2>
        I have double checked my data!
      </h2>
      <Switch
        handleChange={[Function]}
        isOn={false}
        switchId="submit-ready-switch"
      />
    </Section>
  </form>
  <ErrorModal
    handleClose={[Function]}
    show={false}
  />
</Fragment>
`;
